### PR TITLE
Add "mixed" standard form representation

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This adds a new option to the "Standard Form" compiler that returns a mix of inequality and equality constraints.  This representation is particularly useful for solver interfaces.  This PR also resolves an error when collapsing unused variables out of the returned matrices.  The new implementation is also slightly more performant.

## Changes proposed in this PR:
- add a "mixed" representation to the standard form compiler
- correct indexing error when removing unused columns from the `A` matrix and `c` vector 

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/repn/plugins/standard_form.py [hunk 1]
```diff
@@ -372,7 +384,30 @@ def write(self, model):
                     f"model contains a trivially infeasible constraint, '{con.name}'"
                 )
 
-            if slack_form:
+            if mixed_form:
```
Comment (mrmundt) [inline pyomo/repn/plugins/standard_form.py:387]: Is it more likely that a user will want `mixed_form` over `slack_form`?

File: pyomo/repn/tests/test_standard_form.py [hunk 2]
```diff
@@ -222,6 +241,28 @@ def test_alternative_forms(self):
...
+        self.assertEqual(list(map(str, repn.x)), ['x', 'y[0]', 'y[1]', 'y[3]'])
+        self.assertEqual(
+            list(v.bounds for v in repn.x), [(None, None), (0, 10), (-5, 10), (-5, -2)]
+        )
+        ref = np.array(
+            [[1, 0, 2, 0], [0, 0, 1, 4], [0, 1, 6, 0], [0, 1, 6, 0], [1, 1, 0, 0]]
+        )
+        self.assertTrue(np.all(repn.A == ref))
+        print(repn)
+        print(repn.b)
```
Comment (mrmundt) [inline pyomo/repn/tests/test_standard_form.py:260]: Debugging?

File: pyomo/repn/tests/test_standard_form.py [hunk 3]
```diff
@@ -222,6 +241,28 @@ def test_alternative_forms(self):
...
+        self.assertTrue(np.all(repn.b == np.array([3, 5, 6, -3, 8])))
+        self.assertTrue(np.all(repn.c == np.array([[-1, 0, -5, 0], [1, 0, 0, 15]])))
+        # Note that the solution is a mix of inequality and equality constraints
+        # self._verify_solution(soln, repn, False)
```
Comment (mrmundt) [inline pyomo/repn/tests/test_standard_form.py:264]: Debugging?

File: pyomo/repn/plugins/standard_form.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/repn/plugins/standard_form.py:387]: *Users* will probably do the regular form first, and then maybe the slack form.  However, I am working on some (more efficient) solver interfaces (a rewrite of `gurobi_direct`)  that will use the `mixed_form`.

File: pyomo/repn/tests/test_standard_form.py (hunk 3 above)
Comment (jsiirola) [reply pyomo/repn/tests/test_standard_form.py]: Not debugging: a breadcrumb as to why that test is not an exact copy-and-paste of the cases above.  I have expanded the comment.

File: pyomo/repn/tests/test_standard_form.py [hunk 4]
```diff
@@ -222,6 +241,28 @@ def test_alternative_forms(self):
...
+        self.assertEqual(list(map(str, repn.x)), ['x', 'y[0]', 'y[1]', 'y[3]'])
+        self.assertEqual(
+            list(v.bounds for v in repn.x), [(None, None), (0, 10), (-5, 10), (-5, -2)]
+        )
+        ref = np.array(
+            [[1, 0, 2, 0], [0, 0, 1, 4], [0, 1, 6, 0], [0, 1, 6, 0], [1, 1, 0, 0]]
+        )
+        self.assertTrue(np.all(repn.A == ref))
+        self.assertTrue(np.all(repn.b == np.array([3, 5, 6, -3, 8])))
+        self.assertTrue(np.all(repn.c == np.array([[-1, 0, -5, 0], [1, 0, 0, 15]])))
```
Comment (emma58) [inline pyomo/repn/tests/test_standard_form.py:260]: EEEEEK, I didn't notice this before--should we really let a "standard form" have more than one objective? This caused me a moment of terror on the the thought that I would definitely assume `c` if a vector of the same dimension as the number of columns in `A`... The matrix `c` makes sense to me from a writer perspective. From a user/transformation perspective, we're always going to need to be checking that this isn't the case... But I guess we have to do that anyway, so why not here?

**Review by emma58 — APPROVED** [review]
This looks good--I'm excited about it! :) My only big question is that I had a little argument with myself about multiple objectives... I feel like you're probably right, but the idea of more than one objective in "standard form" seems a little terrifying.

**Pull request merged**